internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
 
-	return nil
+func (c *InlineCompleter) Stop() {
+	c.wg.Wait()
 }
 
-type AsyncJobCompleter struct {
+// A default concurrency of 100 seems to perform better a much smaller number
+// like 10, but it's quite dependent on environment (10 and 100 bench almost
+// identically on MBA when it's on battery power). This number should represent
+// our best known default for most use cases, but don't consider its choice to
+// be particularly well informed at this point.
+const asyncCompleterDefaultConcurrency = 100
+
+type AsyncCompleter struct {
 	baseservice.BaseService
+	withSubscribe
 
-	concurrency     uint32
-	exec            PartialExecutor
-	eg              *errgroup.Group
-	subscribeFunc   func(update CompleterJobUpdated)
-	subscribeFuncMu sync.Mutex
+	concurrency int
+	errGroup    *errgroup.Group
+	exec        PartialExecutor
+}
+
+func NewAsyncCompleter(archetype *baseservice.Archetype, exec PartialExecutor) *AsyncCompleter {
+	return newAsyncCompleterWithConcurrency(archetype, exec, asyncCompleterDefaultConcurrency)
 }
 
-func NewAsyncCompleter(archetype *baseservice.Archetype, exec PartialExecutor, concurrency uint32) *AsyncJobCompleter {
-	eg := &errgroup.Group{}
-	// TODO: int concurrency may feel more natural than uint32
-	eg.SetLimit(int(concurrency))
+func newAsyncCompleterWithConcurrency(archetype *baseservice.Archetype, exec PartialExecutor, concurrency int) *AsyncCompleter {
+	errGroup := &errgroup.Group{}
+	errGroup.SetLimit(concurrency)
 
-	return baseservice.Init(archetype, &AsyncJobCompleter{
+	return baseservice.Init(archetype, &AsyncCompleter{
 		exec:        exec,
 		concurrency: concurrency,
-		eg:          eg,
+		errGroup:    errGroup,
 	})
 }
 
-func (c *AsyncJobCompleter) JobSetStateIfRunning(stats *jobstats.JobStatistics, params *riverdriver.JobSetStateIfRunningParams) error {
-	return c.doOperation(stats, func(ctx context.Context) (*rivertype.JobRow, error) {
-		return c.exec.JobSetStateIfRunning(ctx, params)
-	})
-}
-
-func (c *AsyncJobCompleter) doOperation(stats *jobstats.JobStatistics, f func(ctx context.Context) (*rivertype.JobRow, error)) error {
-	c.eg.Go(func() error {
-		start := c.TimeNowUTC()
+func (c *AsyncCompleter) JobSetStateIfRunning(ctx context.Context, stats *jobstats.JobStatistics, params *riverdriver.JobSetStateIfRunningParams) error {
+	// Start clock outside of goroutine so that the time spent blocking waiting
+	// for an errgroup slot is accurately measured.
+	start := c.TimeNowUTC()
 
-		job, err := withRetries(&c.BaseService, f)
+	c.errGroup.Go(func() error {
+		job, err := withRetries(ctx, &c.BaseService, func(ctx context.Context) (*rivertype.JobRow, error) {
+			return c.exec.JobSetStateIfRunning(ctx, params)
+		})
 		if err != nil {
 			return err
 		}
 
 		stats.CompleteDuration = c.TimeNowUTC().Sub(start)
+		c.sendJobToSubscription(job, stats)
+
+		return nil
+	})
+	return nil
+}
+
+func (c *AsyncCompleter) Start(ctx context.Context) error { return nil }
```

Comment:
This context is unused, but do you think it should maybe in some way be a parent to any completion attempts in here? Actually, I'm second-guessing myself in realtime. Even in the case of an aggressive shutdown, we still don't want to shut down until the completer has finished its work. So IMO these completion attempts should not inherit any deadlines whatsoever from parent contexts.